core/node/node_sync/src/batch_transaction_updater/mod.rs

```diff
@@ -0,0 +1,302 @@
+
+        tracing::info!(
+            "Updated finality status for transaction {} ({} for batch {}) from {:?} to {:?}",
+            db_eth_history_tx.tx_hash,
+            db_eth_history_tx.tx_type,
+            batch_number,
+            db_eth_history_tx.eth_tx_finality_status,
+            updated_status
+        );
+
+        Ok(())
+    }
+
+    async fn update_statuses(
+        &self,
+        mut connection: Connection<'_, Core>,
+        to_process: Vec<TxHistory>,
+        l1_block_numbers: L1BlockNumbers,
+    ) -> anyhow::Result<i32> {
+        let mut updated_count = 0;
+
+        tracing::debug!("Checking {} nonfinalized transactions", to_process.len());
+
+        for mut db_eth_tx_history in to_process {
+            // we save receipt here to avoid fetching multiple times
+            let mut receipt: Option<TransactionReceipt> = None;
+
+            // if we didn't see this transaction mined, fetch it
+            let sent_at_block = match db_eth_tx_history.sent_at_block {
+                Some(block) => block,
+                None => {
+                    receipt = self.sl_client.tx_receipt(db_eth_tx_history.tx_hash).await?;
+                    match receipt {
+                        None => {
+                            // transaction was not included, skip. We will try fetching on next iteration.
+                            continue;
+                        }
+                        Some(ref receipt) => {
+                            let sent_at_block: u32 = receipt.block_number.unwrap().as_u32();
+                            db_eth_tx_history.sent_at_block = Some(sent_at_block);
+                            connection
+                                .eth_sender_dal()
+                                .set_sent_at_block(db_eth_tx_history.id, sent_at_block)
+                                .await?;
+                            sent_at_block
+                        }
+                    }
+                }
+            };
+
+            // transaction as included, check if we can potentially update
+            // its finality status. This value is untrusted as seen_at_block may have been cached
+            let Some(finality_update) = l1_block_numbers
+                .get_finality_update(db_eth_tx_history.eth_tx_finality_status, sent_at_block)
+            else {
+                continue;
+            };
+
+            // we can potentially update finality status, but we need to
+            // validate all properties including block number as it may
+            // have been cached
+            let receipt = match receipt {
+                Some(receipt) => receipt,
+                None => {
+                    let Some(receipt) =
+                        self.sl_client.tx_receipt(db_eth_tx_history.tx_hash).await?
+                    else {
+                        tracing::warn!(
+                            "Expected transaction {} to be mined at block {}, but transaction not mined at all on SL",
```

Comment:
Dumb question: How can this happen? Why is this not a persistent error?

Reply:
This happens when there is SL reorg. The transaction may have been first included at block X, but due to a fork (before reaching finalization) it is now on block Y. This does not happen (unless there is some kind of manual block revert) on Gateway, but may happen on L1 for a few blocks. Overall, in principle, its not a disallowed situation for chain to be reorged before finalization.

There are 5-10 reorgs per day on eth mainnet. If the transactions had been included in one of those and we would record the block number, this may get triggered if the block number in receipt changes due to such an reorg. Overall, this should be a very rare occurrence because of how we don't get the transactions on ENs immediately after they are included.